exchanges/coinbase/coinbase.go

```diff
@@ -0,0 +1,1763 @@
+}
+
+// CommitConvertTrade commits a conversion between two currencies, using the trade_id returned from CreateConvertQuote
+func (e *Exchange) CommitConvertTrade(ctx context.Context, tradeID, from, to string) (*ConvertResponse, error) {
+	if tradeID == "" {
+		return nil, errTransactionIDEmpty
+	}
+	if from == "" || to == "" {
+		return nil, errAccountIDEmpty
+	}
+	path := v3Path + convertPath + "/" + tradePath + "/" + tradeID
+	var resp ConvertWrapper
+	return &resp.Trade, e.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodPost, path, nil, convertTradeReqBase{FromAccount: from, ToAccount: to}, true, &resp)
+}
+
+// CreateConvertQuote creates a quote for a conversion between two currencies. The trade_id returned can be used to commit the trade, but that must be done within 10 minutes of the quote's creation
+func (e *Exchange) CreateConvertQuote(ctx context.Context, from, to, userIncentiveID, codeVal string, amount float64) (*ConvertResponse, error) {
+	if from == "" || to == "" {
+		return nil, errAccountIDEmpty
+	}
+	if amount <= 0 {
+		return nil, order.ErrAmountIsInvalid
+	}
+	path := v3Path + convertPath + "/" + quotePath
+	req := convertQuoteReqBase{
+		FromAccount: from,
+		ToAccount:   to,
+		Amount:      amount,
+		Metadata: tradeIncentiveMetadata{
+			UserIncentiveID: userIncentiveID,
+			CodeVal:         codeVal,
+		},
+	}
+	var resp ConvertWrapper
+	return &resp.Trade, e.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodPost, path, nil, req, true, &resp)
+}
+
+// GetConvertTradeByID returns information on a conversion between two currencies
+func (e *Exchange) GetConvertTradeByID(ctx context.Context, tradeID, from, to string) (*ConvertResponse, error) {
+	if tradeID == "" {
+		return nil, errTransactionIDEmpty
+	}
+	if from == "" || to == "" {
+		return nil, errAccountIDEmpty
+	}
+	path := v3Path + convertPath + "/" + tradePath + "/" + tradeID
+	var resp ConvertWrapper
+	return &resp.Trade, e.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, path, nil, convertTradeReqBase{FromAccount: from, ToAccount: to}, true, &resp)
+}
+
+// GetPermissions returns the permissions associated with the API key
+func (e *Exchange) GetPermissions(ctx context.Context) (*PermissionsResponse, error) {
+	var resp PermissionsResponse
+	return &resp, e.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, v3Path+keyPermissionsPath, nil, nil, true, &resp)
+}
+
+// GetTransactionSummary returns a summary of transactions with fee tiers, total volume, and fees
+func (e *Exchange) GetTransactionSummary(ctx context.Context, startDate, endDate time.Time, productVenue, productType, contractExpiryType string) (*TransactionSummary, error) {
+	var params Params
+	params.Values = url.Values{}
+	if err := params.encodeDateRange(startDate, endDate, startDateString, endDateString); err != nil {
```

Comment:
🚧 I quite liked this pattern at first.
But then I realised we are always declaring it, and then calling a method on it, right?
So ... cut out the type and just have a func that returns a straight `url.Values`
You can either make the 2 funcs that we call just return a fresh `url.Values`, or you could pass one in.
The former approach seems like less code everywhere.
All the lines below this just become `params.Set` instead, and everything is clearer.
```suggestion
	params := encodeDateRange(startDate, endDate, startDateString, endDateString)
	if err != nil {
```

Reply:
Yeah, it was set up this way from the start due to the potential of calling both methods within one function, but that hasn't made its way to the current version. I'll make that swap.